### PR TITLE
specification: Segregate object specific flows

### PR DESCRIPTION
**Pull request by nnDarshan**
Specification for segregating all the actions specific
to objects as flows under objects' namespace.

Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

### CODE REVIEW

**Comment by nnDarshan** [comment]
@nthomas-redhat @shtripat @r0h4n @brainfunked  Please review

File: specs/segragate_object_specific_flows.adoc [hunk 1]
```diff
@@ -0,0 +1,295 @@
+===============================
+Segregate object specific flows
```
Comment (shtripat) [inline specs/segragate_object_specific_flows.adoc]: Are we following ascii doc formatting here.
Headings should start with "=", "==" etc..

File: specs/segragate_object_specific_flows.adoc [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+Have all the actions specific to an object as a flow under the objects'
+namespace.
+Example:
+Lets consider the case of volume as an object. In this case the actions specific
+to volume are start, stop, delete etc. All these actions should be defined as
+flows under the volumes' namespace.
+
+sample yaml definition for this:
+
+namespace.tendrl.gluster_integration:
```
Comment (shtripat) [inline specs/segragate_object_specific_flows.adoc:48]: Put the sample under code snippet using ```xyz```

File: specs/segragate_object_specific_flows.adoc [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+          type: String
+        volname:
+          help: "Name of gluster volume"
+          type: String
+      enabled: true
+      value: clusters/$Tendrl_context.cluster_id/Volumes/$Volume.vol_id/
+tendrl_schema_version: 0.3
+
+Note that the flows specific to volume namely delete volume and start volume
+has not dependency on any other object other than volume.
```
Comment (shtripat) [inline specs/segragate_object_specific_flows.adoc]: %s/not/no/g

File: specs/segragate_object_specific_flows.adoc [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+Alternatives
+------------
+
+None
+
+Data model impact
+-----------------
+
+The flows(only object-specific flows) which are currently placed in the global
+nemespace of a particular entity, will be moved under the namespace of the
```
Comment (shtripat) [inline specs/segragate_object_specific_flows.adoc]: %s/nemespace/namespace/g

File: specs/segragate_object_specific_flows.adoc [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+None
+
+Other deployer impact
+---------------------
+
+None
+
+Developer impact
+----------------
```
Comment (shtripat) [inline specs/segragate_object_specific_flows.adoc]: Dont use "-----" and '=====" just like that as they have special meaning in .adoc files

File: specs/segragate_object_specific_flows.adoc [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+-----------------
+
+The flows(only object-specific flows) which are currently placed in the global
+nemespace of a particular entity, will be moved under the namespace of the
+specific object. Example in the "Proposed change" section gives a clear picture.
+
+REST API impact
+---------------
+
+API layer can benefit from this proposed change, as it can now easily figure
```
Comment (r0h4n) [inline specs/segragate_object_specific_flows.adoc:220]: How is the tendrl-api planning to discover these flows under the object's namespace?

**Review by r0h4n — CHANGES_REQUESTED** [review]

File: specs/segragate_object_specific_flows.adoc (hunk 6 above)
Comment (nnDarshan) [reply specs/segragate_object_specific_flows.adoc:220]: @anivargi Could you please provide your inputs

**Comment by nnDarshan** [comment]
@anivargi Please have a look

File: specs/segragate_object_specific_flows.adoc [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+while adding new yaml definitions for flows in the future as well.
+
+== Use Cases
+
+* This specification has to be followed while adding new definition to yaml
+in the future- Developer.
+
+* API layer can figure out the list of actions specific to an object by looking
+at the list of flows under that particular object- API Developer.
+
```
Comment (cybernth) [inline specs/segragate_object_specific_flows.adoc:34]: what about flows which referes multiple objects. It should be as it is now(in global namespece). Please state that explicitly in the document

File: specs/segragate_object_specific_flows.adoc [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+== Testing
+
+Since this is a change in the yaml definition, the testing has to be
+done as part of the yaml validator that we have in the code.
+
+== Documentation Impact
+
+None
+
+== References
```
Comment (cybernth) [inline specs/segragate_object_specific_flows.adoc:280]: You need raise github issues
- on the specification repo(topic issue)
- Raise github issues on all affected reposiories
Refer the github issues in this section of the doc

**Review by cybernth — CHANGES_REQUESTED** [review]

File: specs/segragate_object_specific_flows.adoc (hunk 7 above)
Comment (nnDarshan) [reply specs/segragate_object_specific_flows.adoc:34]: Done

File: specs/segragate_object_specific_flows.adoc (hunk 8 above)
Comment (nnDarshan) [reply specs/segragate_object_specific_flows.adoc:280]: Done

**Comment by nnDarshan** [comment]
@anivargi @brainfunked @shtripat Please review

File: specs/segragate_object_specific_flows.adoc [hunk 9]
```diff
@@ -18,6 +18,7 @@ By defining the objects specific actions as flows under objects' namespace,
 we can easily relate the flows with the corresponding objects'. It will be easy
 for the API to figure out the actions that can be performed on the object. Care
 must be taken to see that such flows are not dependent on any other objects'.
+The flows which are dependent on multiple objects should be at the global leval
```
Comment (shtripat) [inline specs/segragate_object_specific_flows.adoc]: %s/leval/level/g

File: specs/segragate_object_specific_flows.adoc [hunk 10]
```diff
@@ -201,6 +202,9 @@ tendrl_schema_version: 0.3
 Note that the flows specific to volume namely delete volume and start volume
 has no dependency on any other object other than volume.
 
+The flows that are dependent on multiple objects should be retained at the
+global leval.
```
Comment (cybernth) [inline specs/segragate_object_specific_flows.adoc]: %s/leval/level/g
Here as well :)

**Review by cybernth — COMMENTED** [review]
Looks good to me. I wll approve is you make this minor change

File: specs/segragate_object_specific_flows.adoc (hunk 9 above)
Comment (nnDarshan) [reply specs/segragate_object_specific_flows.adoc]: Done

File: specs/segragate_object_specific_flows.adoc (hunk 10 above)
Comment (nnDarshan) [reply specs/segragate_object_specific_flows.adoc]: Done

**Review by shtripat — COMMENTED** [review]
Looks good to me

**Comment by anivargi** [comment]
Just to understand this clearly, if a object is not dependent on any other object, then the flow will be inside the objects, else it will be global?

CreateVolume is global as of now which is a creating a volume resource, but everything else is working on a existing resource, if we do that kind of bifurcation then it also makes sense, but the above assumption does not make any sense. If we have to refer a different object inside the CreateVolume flow then it will move to inside the Volume namespace?

I would rather propose everything to do with a Volume namespace to be inside Volume.

**Comment by r0h4n** [comment]
@anivargi 

- If a flow deals with only one object via its "inputs" and "outputs", then it belongs to that object's namespace

- If a flow deals with multiple objects via its "inputs" and "outputs", then the flow belongs to the global namespace

makes sense?

**Review by r0h4n — APPROVED** [review]

**Comment by brainfunked** [comment]
Against https://github.com/Tendrl/specifications/issues/34

**Pull request merged**